libyul/backends/wasm/BinaryTransform.cpp

```diff
@@ -332,12 +361,18 @@ bytes BinaryTransform::operator()(BuiltinCall const& _call)
 	{
 		yulAssert(builtins.count(_call.functionName), "Builtin " + _call.functionName + " not found");
 		bytes ret = std::move(args) + toBytes(builtins.at(_call.functionName));
+
 		if (
 			_call.functionName.find(".load") != string::npos ||
 			_call.functionName.find(".store") != string::npos
 		)
+			// Just a heuristic to detect someone forgetting to specify alignment when adding new a builtin.
+			yulAssert(builtinAlignment.count(_call.functionName) == 1, "");
+
+		if (builtinAlignment.count(_call.functionName) == 1)
 			// alignment and offset
-			ret += bytes{{3, 0}};
```

Comment:
I had a much simpler version locally: {0, 0}.

It is safe setting it to a lower alignment as the alignment is only a hint for the VM, but most of them don't care.

Reply:
I thought about that but since it was originally set to 8 (presumably for the 64-bit types) I assumed that 0 would not cut it :)

Reply:
It was set to 2^3=8 and I propose we just set it to 2^0=1 for the sake of simplicity. 

But since you implemented all this anyway, I'm fine merging it.

Reply:
Ah, sorry, I meant 1 (2^0), not 0.

> But since you implemented all this anyway, I'm fine merging it.

Now that I think of it - what happens if the hint is wrong? I.e. we say it's aligned to 2/4/8 bytes and it's not?

Reply:
It is only a hint to the virtual machine to optimise access. Validation rules allow it to be lower alignment than the native size, e.g. byte-aligned is always fine.

Interpreters ignore this. JITs/AOTs can take it into account to generate more efficient code.

In the blockchain context we do not expect anything, but interpreters, so likely this will not have any effect. I also do not have any benchmark numbers how that affects AOTs.

For completeness it is nicer to have it.

Reply:
OK. Let's go with this PR then. I need to change base again in that case.

Reply:
While thinking about how I would deal with alignment hints if I were to write an AOT: the alignment indicated the actual offsets in memory are aligned. If the hint is invalid, that is a miss and it will be more expensive cycle wise (the VMs have to be aware of this since there is no validation).

Assuming the wasm memory is allocated as aligned to 8 bytes, having the offsets within that linear memory aligned is the onus of the wasm binary, aka. the solidity compiler.

I doubt that we try to align it, so in practice we give invalid hints most of the time.

My conclusions is to hint at byte alignment, unless we plan to control the memory layout accordingly.

Reply:
Right. I changed it to 1-byte alignment for everything now.